dali/pipeline/operator/builtin/merge.cc

```diff
@@ -0,0 +1,173 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+
+#include "dali/core/common.h"
+#include "dali/core/util.h"
+#include "dali/pipeline/data/backend.h"
+#include "dali/pipeline/data/types.h"
+#include "dali/pipeline/operator/builtin/merge.h"
+#include "dali/pipeline/operator/builtin/split_merge.h"
+
+namespace dali {
+
+
+template <typename Backend>
+bool Merge<Backend>::SetupImpl(std::vector<OutputDesc> &output_desc, const Workspace &ws) {
+  input_sample_count_ = 0;
+  int nonzero_input_sample_idx = -1;
+  for (int input_category_idx = 0; input_category_idx < kMaxCategories; input_category_idx++) {
+    const auto &input = ws.template Input<Backend>(input_category_idx);
+    input_sample_count_ += input.num_samples();
+    // We could do additional error checking for input consistency, but the SetSample already does
+    // the same.
+  }
+
+  const auto &predicate = ws.ArgumentInput("predicate");
+  DALI_ENFORCE(
+      input_sample_count_ == predicate.num_samples(),
+      make_string("Merge description must cover whole input, got ", input_sample_count_,
+                  " input samples and ", predicate.num_samples(), " elements denoting the merge."));
+  DALI_ENFORCE(predicate.shape().sample_dim() == 0, "Only scalar indexing is supported.");
+  return false;
+}
+
+
+template <typename Backend>
+void Merge<Backend>::RunImpl(Workspace &ws) {
+  auto &output = ws.template Output<Backend>(0);
+  const auto &predicate = ws.ArgumentInput("predicate");
+  auto sample_idx_in_input = uniform_array<kMaxCategories>(0);
+
+  WriteTestsDiagnostics(ws);
+
+  if (!pinned_) {
+    // We produce pinned data if the executor said so
+    pinned_ = output.is_pinned();
+    // TODO(klecki): We keep pinedness if it was passed to us:
+    for (int input_category = 0; input_category < kMaxCategories; input_category++) {
+      pinned_ = *pinned_ || ws.template Input<Backend>(input_category).is_pinned();
+    }
+  }
+
+  if (*pinned_ || std::is_same_v<Backend, GPUBackend>) {
+    cudaGetDevice(&device_id_);
+  } else {
+    device_id_ = CPU_ONLY_DEVICE_ID;
+  }
+
+  // We propagate views only, so just don't care about what is here and reset
+  output.Reset();
+  for (int input_category = 0; input_category < kMaxCategories; input_category++) {
+    const auto &input = ws.template Input<Backend>(input_category);
+    if (input.num_samples() > 0) {
+      output.set_type(input.type());
+      output.set_sample_dim(input.shape().sample_dim());
+      output.SetLayout(input.GetLayout());
```

Comment:
I think you should check in SetupImpl that all your inputs have the same type, dimensionality and layout.

Reply:
SetSample already does it, but I add more suited error messaging.